types/block.go

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright 2018 The CovenantSQL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package types
+
+import (
+	"time"
+
+	"github.com/CovenantSQL/CovenantSQL/crypto"
+	ca "github.com/CovenantSQL/CovenantSQL/crypto/asymmetric"
+	"github.com/CovenantSQL/CovenantSQL/crypto/hash"
+	"github.com/CovenantSQL/CovenantSQL/crypto/kms"
+	"github.com/CovenantSQL/CovenantSQL/merkle"
+	"github.com/CovenantSQL/CovenantSQL/proto"
+	"github.com/CovenantSQL/CovenantSQL/utils/log"
+)
+
+//go:generate hsp
+
+// Header is a block header.
+type Header struct {
+	Version     int32
+	Producer    proto.NodeID
+	GenesisHash hash.Hash
+	ParentHash  hash.Hash
+	MerkleRoot  hash.Hash
+	Timestamp   time.Time
+}
+
+// SignedHeader is block header along with its producer signature.
+type SignedHeader struct {
+	Header
+	HSV crypto.DefaultHashSignVerifierImpl
+}
+
+func (s *SignedHeader) Sign(signer *ca.PrivateKey) error {
+	return s.HSV.Sign(&s.Header, signer)
+}
+
+// Verify verifies the signature of the signed header.
+func (s *SignedHeader) Verify() error {
+	return s.HSV.Verify(&s.Header)
+}
+
+// VerifyAsGenesis verifies the signed header as a genesis block header.
+func (s *SignedHeader) VerifyAsGenesis() (err error) {
+	var pk *ca.PublicKey
+	log.WithFields(log.Fields{
+		"producer": s.Producer,
+		"root":     s.GenesisHash.String(),
+		"parent":   s.ParentHash.String(),
+		"merkle":   s.MerkleRoot.String(),
+		"block":    s.HSV.Hash().String(),
+	}).Debug("Verifying genesis block header")
+	if pk, err = kms.GetPublicKey(s.Producer); err != nil {
+		return
+	}
+	if !pk.IsEqual(s.HSV.Signee) {
+		err = ErrNodePublicKeyNotMatch
+		return
+	}
+	return s.Verify()
+}
+
+type QueryAsTx struct {
```

Comment:
**[golint]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
exported type QueryAsTx should have comment or be unexported